go-controller/pkg/cni/udn/primary_network.go

```diff
@@ -107,53 +124,119 @@ func (p *UserDefinedPrimaryNetwork) ensure(namespace string, annotations map[str
 
 	// If default network is the primary there is nothing else to do
 	if annotation.Role == types.NetworkRolePrimary {
-		return nil
+		return true, nil
 	}
 
-	if err := p.ensureAnnotation(annotations); err != nil {
-		return fmt.Errorf("failed looking for primary network annotation: %w", err)
+	ready, err := p.ensureAnnotation(pod)
+	if !ready || err != nil {
+		return ready, err
 	}
-	if err := p.ensureActiveNetwork(namespace); err != nil {
-		return fmt.Errorf("failed ensuring primary network for namespace: %q: %w", namespace, err)
+	ready, err = p.ensureActiveNetwork(pod.Namespace)
+	if !ready || err != nil {
+		return ready, err
 	}
-	return nil
+	return p.ensureNetworkDevice(pod)
 }
 
-func (p *UserDefinedPrimaryNetwork) ensureActiveNetwork(namespace string) error {
+func (p *UserDefinedPrimaryNetwork) ensureActiveNetwork(namespace string) (bool, error) {
 	if p.activeNetwork != nil {
-		return nil
+		return true, nil
 	}
 	activeNetwork, err := p.networkManager.GetActiveNetworkForNamespace(namespace)
 	if err != nil {
-		return err
+		return false, err
 	}
 	if activeNetwork.IsDefault() {
-		return fmt.Errorf("missing primary user defined network NAD for namespace '%s'", namespace)
+		return false, fmt.Errorf("missing primary user defined network NAD for namespace '%s'", namespace)
 	}
+
 	p.activeNetwork = activeNetwork
-	return nil
+	return true, nil
 }
 
-func (p *UserDefinedPrimaryNetwork) ensureAnnotation(annotations map[string]string) error {
+func (p *UserDefinedPrimaryNetwork) ensureAnnotation(pod *corev1.Pod) (bool, error) {
 	if p.annotation != nil {
-		return nil
+		return true, nil
 	}
+	annotations := pod.GetAnnotations()
 	podNetworks, err := util.UnmarshalPodAnnotationAllNetworks(annotations)
 	if err != nil {
-		return err
+		return false, err
 	}
 	for nadName, podNetwork := range podNetworks {
 		if podNetwork.Role != types.NetworkRolePrimary {
 			continue
 		}
-		p.annotation, err = util.UnmarshalPodAnnotation(annotations, nadName)
+		annotation, err := util.UnmarshalPodAnnotation(annotations, nadName)
+		if err != nil {
+			if util.IsAnnotationNotSetError(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		// To support for non VFIO devices like SRIOV, get the primary UDN's resource name
+		nad, err := p.nadLister.NetworkAttachmentDefinitions(pod.Namespace).Get(strings.TrimPrefix(nadName, pod.Namespace+"/"))
 		if err != nil {
-			return err
+			return false, fmt.Errorf("failed to get primary UDN's network-attachment-definition %s: %v", nadName, err)
 		}
-		break
+		p.annotation = annotation
+		p.nadName = nadName
+		p.resourceName = nad.Annotations[resourceNameAnnot]
+		klog.Infof("Primary UDN for pod %s/%s NAD %s resource %s", pod.Namespace, pod.Name, p.nadName, p.resourceName)
+
+		return true, nil
+	}
+	return false, nil
+}
+
+func (p *UserDefinedPrimaryNetwork) ensureNetworkDevice(pod *corev1.Pod) (bool, error) {
```

Comment:
this function doesn't need to return (bool, error), just error. There is no case where it can return false nil. No error means ready.

Reply:
done